File: support/jolie-gwt/src/joliex/gwt/client/Value.java
/***************************************************************************
 *   Copyright (C) by Fabrizio Montesi                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Library General Public License as       *
 *   published by the Free Software Foundation; either version 2 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU Library General Public     *
 *   License along with this program; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   For details about the authors of this software, see the AUTHORS file. *
 ***************************************************************************/


package joliex.gwt.client;

import com.google.gwt.user.client.rpc.IsSerializable;
import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

public class Value implements Serializable, IsSerializable
{
	public enum Type implements IsSerializable {
		UNDEFINED, STRING, INT, DOUBLE, LONG, BOOLEAN, BYTEARRAY
	}

	// This field must not be final or GWT will not serialize it correctly
	private Map< String, ValueVector > children = new HashMap< String, ValueVector >();

	private String valueObject = null;
	private Type type = Type.UNDEFINED;
	
	public Value()
	{}
	
	public Value( String value )
	{
		setValue( value );
	}
	
	public Value( Integer value )
	{
		setValue( value );
	}
	
	public Value( Double value )
	{
		setValue( value );
	}
        // Added by Balint Maschio 
	public Value( Long value )
	{
		setValue( value );
	}
        public Value( Boolean value )
	{
		setValue( value );
	}

	public Value( ByteArray value )
	{
		setValue( value );
	}

	public boolean isString()
	{
		return type == Type.STRING;
	}
	
	public boolean isInt()
	{
		return type == Type.INT;
	}
	
	public boolean isDouble()
	{
		return type == Type.DOUBLE;
	}
        
	// Added by Balint Maschio
	public boolean isLong()
	{
		return type == Type.LONG;
	}

	// Added by Balint Maschio
	public boolean isBool()
	{
		return type == Type.BOOLEAN;
	}

	public boolean isByteArray()
	{
		return type == Type.BYTEARRAY;
	}

	public boolean isDefined()
	{
		return type != Type.UNDEFINED;
	}
	
	public ValueVector getChildren( String id )
	{
		ValueVector v = children.get( id );
		if ( v == null ) {
			v = new ValueVector();
			children.put( id, v );
		}
	
		return v;	
	}
	
	public boolean hasChildren()
	{
		return !children.isEmpty();
	}
	
	public boolean hasChildren( String id )
	{
		return children.get( id ) != null;
	}

	public void deepCopy( Value otherValue )
	{
		valueObject = otherValue.valueObject;
                type = otherValue.type;
		ValueVector myVector;
		Value myValue;
		for( Entry< String, ValueVector > entry : otherValue.children.entrySet() ) {
			myVector = new ValueVector();
			for( Value v : entry.getValue() ) {
				myValue = new Value();
				myValue.deepCopy( v );
				myVector.add( v );
			}
			children.put( entry.getKey(), myVector );
		}
	}

	public String strValue()
	{
		if ( valueObject == null )
			return new String();
		return valueObject;
	}

	public int intValue()
	{
		if ( valueObject == null )
			return 0;
		return Integer.valueOf( valueObject );
	}

	public double doubleValue()
	{
		if ( valueObject == null )
			return 0.0;
		return Double.valueOf( valueObject );
	}

	// Added by Balint Maschio
	public long longValue()
	{
		if ( valueObject == null )
			return 0L;
		return Long.valueOf( valueObject );
	}

	public boolean boolValue()
	{
		if ( valueObject == null )
			return false;
		return Boolean.valueOf( valueObject );
	}

        public ByteArray byteArrayValue() {
                
                ByteArray r = null;
		if ( valueObject == null ) {
                        byte[] resp = new byte[0];
			return new ByteArray( resp );
		} else {
                    char[] chars = valueObject.toCharArray();
                    byte[] byteArrayToReturn= new byte[chars.length * 2 ];  //bytes per char = 2
                    for (int i = 0; i < chars.length; i++)
                    {
                        for (int j = 0; j < 2; j++)
                            byteArrayToReturn[i * 2 + j] = (byte) (chars[i] >>> (8 * (1 - j)));
                    }
                    
                    return new ByteArray( byteArrayToReturn );
                }
        }
	
	public Value getNewChild( String childId )
	{
		ValueVector vec = getChildren( childId );
		Value retVal = new Value();
		vec.add( retVal );
		
		return retVal;
	}
	
	public Map< String, ValueVector > children()
	{
		return children;
	}
	
	public Value getFirstChild( String id )
	{
		return getChildren( id ).first();
	}
	
	public final void setValue( String obj )
	{
		valueObject = obj;
		type = Type.STRING;
	}

	public final void setValue( Integer obj )
	{
		valueObject = obj.toString();
		type = Type.INT;
	}
	
	public final void setValue( Double obj )
	{
		valueObject = obj.toString();
		type = Type.DOUBLE;
                
	}
	
	// Added by Balint Maschio
	public final void setValue( Long obj )
	{
		valueObject = obj.toString();
		type = Type.LONG;
                
	}
        
	public final void setValue( Boolean obj )
	{
		valueObject = obj.toString();
		type = Type.BOOLEAN;
	}
        
	public final void setValue( ByteArray obj )
	{
		valueObject = obj.toString();
		type = Type.BYTEARRAY;
	}
        
}

File: c-sharp_library/JolieLib/Jolie/runtime/Value.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace Jolie.runtime
{
    public class Value : ISerializable
    {
        //[Serializable]
        public enum Type
        {
            UNDEFINED, STRING, INT, DOUBLE, LONG, BOOLEAN, BYTEARRAY
        }

        private Dictionary<String, ValueVector> children = new Dictionary<String, ValueVector>();

        private String valueObject = null;
        private Type type = Type.UNDEFINED;

        public Boolean HasChildren { get { return children.Count > 0; } }

        public Dictionary<String, ValueVector> Children { get { return children == null ? new Dictionary<String, ValueVector>() : children; } }

        #region Identifiers

        public Type GetValueType()
        {
            if (IsString) return Type.STRING;
            else if (IsInt) return Type.INT;
            else if (IsDouble) return Type.DOUBLE;
            else if (IsLong) return Type.LONG;
            else if (IsBool) return Type.BOOLEAN;
            else if (IsByteArray) return Type.BYTEARRAY;
            else return Type.UNDEFINED;
        }

        public Boolean IsString { get { return type == Type.STRING; } }

        public Boolean IsInt { get { return type == Type.INT; } }

        public Boolean IsDouble { get { return type == Type.DOUBLE; } }

        public Boolean IsLong { get { return type == Type.LONG; } }

        public Boolean IsBool { get { return type == Type.BOOLEAN; } }

        public Boolean IsDefined { get { return type != Type.UNDEFINED; } }

        public Boolean IsByteArray { get { return type == Type.BYTEARRAY; } }

        #endregion

        #region Constructors

        public Value() { }

        public Value(ByteArray value)
        {
            SetValue(value);
        }

        public Value(String value)
        {
            SetValue(value);
        }

        public Value(Int32 value)
        {
            SetValue(value);
        }

        public Value(Double value)
        {
            SetValue(value);
        }

        // Added by Balint Maschio 
        public Value(Int64 value)
        {
            SetValue(value);
        }

        public Value(Boolean value)
        {
            SetValue(value);
        }

        #endregion

        #region Value methods

        public int IntValue
        {
            get
            {
                if (valueObject == null) return 0;
                return Int32.Parse(valueObject);
            }
        }

        public double DoubleValue
        {
            get
            {
                if (valueObject == null) return 0;
                return Double.Parse(valueObject);
            }
        }

        public String StrValue
        {
            get
            {
                if (valueObject == null) return "";
                return valueObject.ToString();
            }
        }

        // Added by Balint Maschio
        public long LongValue
        {
            get
            {
                if (valueObject == null) return 0;
                return long.Parse(valueObject);
            }
        }

        public Boolean BoolValue
        {
            get
            {
                if (valueObject == null) return false;
                return Boolean.Parse(valueObject);
            }
        }

        public ByteArray ByteArrayValue
        {
            get
            {
                ByteArray r = null;
                if (valueObject == null)
                {
                    byte[] resp = new byte[0];
                    return new ByteArray(resp);
                }
                else
                {
                    char[] chars = valueObject.ToCharArray();
                    byte[] byteArrayToReturn = new byte[chars.Length * 2];  //bytes per char = 2
                    for (int i = 0; i < chars.Length; i++)
                    {
                        for (int j = 0; j < 2; j++)
                            byteArrayToReturn[i * 2 + j] = (byte)(chars[i] >> (8 * (1 - j))); // OBS, original >>>
                    }
                    return new ByteArray(byteArrayToReturn);
                }
            }
        }

        #endregion

        public ValueVector GetChildren(String id)
        {
            ValueVector v = Children[id];
            if (v == null)
            {
                v = new ValueVector();
                children.Add(id, v);
            }
            return v;
        }

        public Boolean hasChildren(String id)
        {
            return children[id] != null;
        }

        public void deepCopy(Value otherValue)
        {
            valueObject = otherValue.valueObject;
            type = otherValue.type;
            ValueVector myVector;
            Value myValue;
            foreach (KeyValuePair<String, ValueVector> entry in otherValue.children)
            {
                myVector = new ValueVector();
                foreach (Value v in entry.Value)
                {
                    myValue = new Value();
                    myValue.deepCopy(v);
                    myVector.add(v);
                }
                children.Add(entry.Key, myVector);
            }
        }

        public Value GetNewChild(String childId)
        {
            ValueVector vec = GetChildren(childId);
            Value retVal = new Value();
            vec.add(retVal);
            return retVal;
        }

        public Value GetFirstChild(String id)
        {
            return GetChildren(id).First();
        }

        #region Set value methods

        public void SetValue(String obj)
        {
            valueObject = obj;
            type = Type.STRING;
        }

        public void SetValue(Int32 obj)
        {
            valueObject = obj.ToString();
            type = Type.INT;
        }

        public void SetValue(Double obj)
        {
            valueObject = obj.ToString();
            type = Type.DOUBLE;
        }

        public void SetValue(long obj)
        {
            valueObject = obj.ToString();
            type = Type.LONG;
        }

        public void SetValue(Boolean obj)
        {
            valueObject = obj.ToString();
            type = Type.BOOLEAN;
        }

        public void SetValue(ByteArray obj)
        {
            valueObject = obj.ToString();
            type = Type.BYTEARRAY;
        }

        #endregion

        public Boolean isUsedInCorrelation()
        {
            return false;
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            throw new NotImplementedException();
        }
    }
}
